### PR TITLE
Generalize linters support #154

### PR DESCRIPTION
**Pull request by gal432**

### CODE REVIEW

File: lms/lmstests/sandbox/linters/defines.py [hunk 1]
```diff
@@ -141,3 +141,16 @@
     'W503',  # line break after operator
     'W605',  # invalid escape sequence
 )
+
+
+# In case you want to skip an error message in VNU linter, you should write this here.
+# Note that the mapping is based on the whole error message, and not code like in flake8
+# Since there are no 'error codes' in VNU
+VNU_SKIP_ERROR_MESSAGES = (
```
Comment (yammesicka) [inline lms/lmstests/sandbox/linters/defines.py:149]: Maybe type hinting here?

File: lms/lmsdb/models.py [hunk 2]
```diff
@@ -553,6 +553,10 @@ def is_duplicate(
     def create_hash(content: Union[str, bytes], *args, **kwargs) -> str:
         return hashing.by_content(content, *args, **kwargs)
 
+    @property
+    def suffix(self) -> str:
+        return self.path.split('.')[-1]
```
Comment (yammesicka) [inline lms/lmsdb/models.py:558]:
```python
filename, __, ext = self.path.rpartition('.')
return ext if filename else ''
```

File: lms/lmstests/sandbox/linters/services.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        for linter in cls.__subclasses__():
+            if linter.match_to_file_suffix(file_suffix):
+                return linter(
+                    logger=logger,
+                    code=code,
+                    file_suffix=file_suffix,
+                    solution_file_id=solution_file_id,
+                )
+
+    def get_error_text(self, error: LinterError):
```
Comment (yammesicka) [inline lms/lmstests/sandbox/linters/services.py:64]: Return type

File: lms/lmstests/sandbox/linters/services.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        raise NotImplementedError
+
+    @staticmethod
+    def match_to_file_suffix(file_suffix: str) -> bool:
+        raise NotImplementedError
+
+    def initialize(self):
+        """
+        For special initializing code, please write it here
+        """
```
Comment (yammesicka) [inline lms/lmstests/sandbox/linters/services.py:45]: raise NotImplementedError()

File: lms/lmstests/sandbox/linters/services.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+                line_number=error.line_number,
+                column=error.column,
+                text=custom_text,
+                physical_line=error.physical_line,
+                solution_file_id=self._solution_file_id,
+            ))
+        return response
+
+
+class PythonLinter(BaseLinter):
```
Comment (yammesicka) [inline lms/lmstests/sandbox/linters/services.py:92]: Maybe a separate file?

File: lms/lmstests/sandbox/linters/services.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+                    *result, solution_file_id=self._solution_file_id)
+                if response.error_code in defines.FLAKE_SKIP_ERRORS:
+                    self._logger.info(
+                        'Skipping error %s on line %s to solution file %s',
+                        response.error_code,
+                        response.line_number,
+                        self._solution_file_id,
+                    )
+                else:
+                    errors.append(response)
```
Comment (yammesicka) [inline lms/lmstests/sandbox/linters/services.py:130]: Use yield instead

File: lms/lmstests/sandbox/linters/services.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+                    'vnu',
+                    '--format',
+                    'json',
+                    '--also-check-css',
+                    temp_file.name,
+                ),
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+            )
+            output_buffer, error_buffer = process.communicate()
```
Comment (yammesicka) [inline lms/lmstests/sandbox/linters/services.py:167]: You can close the file here

File: lms/lmstests/sandbox/linters/services.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+                    )
+                    if response.text in defines.VNU_SKIP_ERROR_MESSAGES:
+                        self._logger.info(
+                            'Skipping error %s on line %s to solution file %s',
+                            response.error_code,
+                            response.line_number,
+                            self._solution_file_id,
+                        )
+                    else:
+                        errors.append(response)
```
Comment (yammesicka) [inline lms/lmstests/sandbox/linters/services.py:190]: yield

File: lms/lmstests/sandbox/linters/services.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+                    '--also-check-css',
+                    temp_file.name,
+                ),
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+            )
+            output_buffer, error_buffer = process.communicate()
+            results = json.loads(error_buffer)
+            for result in results['messages']:
+                try:
```
Comment (yammesicka) [inline lms/lmstests/sandbox/linters/services.py:170]: Let's focus the try/except to lines 174-181

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: lms/lmstests/sandbox/linters/services.py (hunk 4 above)
Comment (gal432) [reply lms/lmstests/sandbox/linters/services.py:45]: Why? I don't need it in all places

**Review by yammesicka — APPROVED** [review]

**Pull request merged**